go/oasis-test-runner/scenario/e2e/runtime/runtime.go

```diff
@@ -311,91 +288,76 @@ func (sc *runtimeImpl) resolveRuntimeBinary(runtimeBinary string, tee node.TEEHa
+		return fmt.Errorf("failed to initialize network: %w", err)
 	}
 
-	d, err := childEnv.NewSubDir("client")
-	if err != nil {
-		return nil, err
-	}
+	return sc.startTestClientOnly(ctx, childEnv)
+}
 
-	w, err := d.NewLogWriter("client.log")
-	if err != nil {
-		return nil, err
+func (sc *runtimeImpl) startTestClientOnly(ctx context.Context, childEnv *env.Env) error {
+	if err := sc.testClient.Init(sc); err != nil {
+		return fmt.Errorf("failed to initialize test client: %w", err)
 	}
 
-	args := []string{
-		"--node-address", "unix:" + clients[0].SocketPath(),
-		"--runtime-id", runtimeID.String(),
+	if err := sc.testClient.Start(ctx, childEnv); err != nil {
+		return fmt.Errorf("failed to start test client: %w", err)
 	}
-	args = append(args, sc.clientArgs...)
 
-	binary := sc.resolveClientBinary(sc.clientBinary)
-	cmd := exec.Command(binary, args...)
-	cmd.SysProcAttr = env.CmdAttrs
-	cmd.Stdout = w
-	cmd.Stderr = w
+	return nil
+}
 
-	sc.Logger.Info("launching client",
-		"binary", binary,
-		"args", strings.Join(args, " "),
-	)
+func (sc *runtimeImpl) waitTestClientOnly() error {
+	return sc.testClient.Wait()
+}
 
-	if err = cmd.Start(); err != nil {
-		return nil, fmt.Errorf("scenario/e2e: failed to start client: %w", err)
-	}
+func (sc *runtimeImpl) checkTestClientLogs() error {
+	// Wait for logs to be fully processed before checking them. When
+	// the client exits very quickly the log watchers may not have
+	// processed the relevant logs yet.
+	//
+	// TODO: Find a better way to synchronize log watchers.
+	time.Sleep(1 * time.Second)
 
-	return cmd, nil
+	return sc.Net.CheckLogWatchers()
 }
 
-func (sc *runtimeImpl) waitClient(childEnv *env.Env, cmd *exec.Cmd, clientErrCh <-chan error) error {
-	var err error
-	select {
-	case err = <-sc.Net.Errors():
-		_ = cmd.Process.Kill()
-	case err = <-clientErrCh:
-	}
-	if err != nil {
+func (sc *runtimeImpl) waitTestClient() error {
+	if err := sc.waitTestClientOnly(); err != nil {
 		return err
 	}
-
-	return nil
+	return sc.checkTestClientLogs()
 }
 
-func (sc *runtimeImpl) wait(childEnv *env.Env, cmd *exec.Cmd, clientErrCh <-chan error) error {
-	if err := sc.waitClient(childEnv, cmd, clientErrCh); err != nil {
+func (sc *runtimeImpl) Run(childEnv *env.Env) error {
+	ctx := context.Background()
+	if err := sc.startNetworkAndTestClient(ctx, childEnv); err != nil {
 		return err
 	}
-
-	// Wait for logs to be fully processed before checking them. When the client exits very quickly
-	// the log watchers may not have processed the relevant logs yet.
-	// TODO: Find a better way to synchronize log watchers.
-	time.Sleep(1 * time.Second)
-
-	return sc.Net.CheckLogWatchers()
+	return sc.waitTestClient()
 }
 
-func (sc *runtimeImpl) Run(childEnv *env.Env) error {
-	clientErrCh, cmd, err := sc.start(childEnv)
+func (sc *runtimeImpl) submitNakedRuntimeTx(ctx context.Context, id common.Namespace, method string) (cbor.RawMessage, error) {
```

Comment:
These could move to the test client impls?

Reply:
No, not everything uses a test client.